---
title: "Introduction to edgeR GLMs"
author: Daniel Neves and Daniel Sobral
date: "October 10, 2018"
output: 
  html_document: 
    keep_md: yes
---

Here we demonstrate the use of *edgeR* to perform a differential expression analysis using data from Tuch *et al.* ([PLOS](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0009317)) as detailed in the edgeR manual.

The data is a set of RNA-seq samples of oral squamous cell carcinomas and matched normal tissue from three patients that were previously quantified into raw counts. 

We will use *edgeR* to do a differential expression analysis of Tumor vs Non-Tumor samples. We will start with a simple pairwise comparison of the Tumor and Non-Tumor samples, and then repeat the analysis adding the patient pairing information to the model design.

# Load the count data

We start by importing the counts table into R using the `read.delim` function. Other functions to import tables include `read.table` and `read.csv`. We also specify that the values in the tables are separated by a TAB. You can type `?read.delim` in the R console to display the documentation of the function.

```{r}
rawdata <- read.delim("edgeR_example1_Tuch.tab", sep = "\t")
```

To check that the data was loaded properly we can use functions such as `head` (to displays the first lines of the table), `dim` (to display the dimensions of the table) and `summary` (to display summary statistics for each column). In RStudio you can also type `View(rawdata)` to view the full table on a separate window.

```{r}
head(rawdata)
dim(rawdata)
summary(rawdata)
```

For convenience, we separate the table in two: one containing the counts for all samples (columns 2 to 7), and another containing only the list of gene names (column 1).

```{r}
rawcounts <- rawdata[, 2:7]
genes <- rawdata[, 1]
```

# Simple pairwise differential expression analysis with edgeR GLMs

We need to import edgeR into the R environment.

```{r}
library(edgeR)
```

We start by telling edgeR where our raw counts are, and calculate normalization factors.

```{r}
y <- DGEList(counts=rawcounts, genes=genes)
y <- calcNormFactors(y)

y$samples
```

After normalization, we can now produce a Multidimensional Scaling Plot (MDS) using the function `plotMDS`. This visualization, a type of dimensional reduction technique, places the samples on a plane such that the distance between samples approximates the typical log2 fold-changes between them.

```{r}
plotMDS(y)
```

We now define the design of our comparison. We want to compare Tumor to Non-Tumor samples. So we first create a variable indicating which samples are from normal (N) or tumor (T) tissue. Then we define the design for the genewise linear models. Here, the `~ Tissue` design is equivalent to a simple pairwise test of Tumor vs Non-tumor (i.e. the model only takes into account the originating tissue).

```{r}
Tissue <- factor(c("N","T","N","T","N","T"))

design <- model.matrix(~ Tissue)
rownames(design) <- colnames(y)

design
```

Next we use this design to conduct the test of differential expression. In *edgeR*, this is done in 3 steps: estimation of the negative binomial dispersions (`estimateDisp`), fitting of the negative binomial model to the count data (`glmFit`) and hypothesis testing (`glmLRT`).

```{r}
y <- estimateDisp(y, design, robust=TRUE)
fit <- glmFit(y, design)
lrt <- glmLRT(fit)
```

We now check how many genes were differentially expressed.

```{r}
summary(decideTestsDGE(lrt))
```

In *edgeR* we make an MA-plot with the `plotMD` function. Up-regulated genes are indicated in red, and down-regulated genes are indicated in blue. The horizontal lines indicate 2x fold-changes.

```{r}
plotMD(lrt)
abline(h=c(-1, 1), col="blue")
```

We can retrieve a table with all the results of differential expression using the `topTags` function. We also save it to a file so we can latter open it in Excel.

```{r}
result <- as.data.frame(topTags(lrt, n = nrow(rawcounts)))

head(result)

write.table(result, file = "edgeR_Tuch_Tumor_vs_NonTumor.csv", sep="\t", row.names = FALSE)
```

# A more complex design: adding patient pairing information

Recall that tumor and non-samples were collected from 3 patients. Until now we have ignored this information in our design. Here we repeat the analysis by adding the sample pairing information to our model design, that will allow us to adjust for differences between patients.

For this we only have to change the design definition. We create a new `Patient` variable, and then include it as a blocking factor in the GLM design. 

```{r}
Patient <- factor(c(8, 8, 33, 33, 51, 51))
Tissue <- factor(c("N","T","N","T","N","T"))

design <- model.matrix(~ Patient + Tissue)
rownames(design) <- colnames(y)

design

y <- DGEList(counts=rawcounts, genes=genes)
y <- calcNormFactors(y)
y <- estimateDisp(y, design, robust=TRUE)
fit <- glmFit(y, design)
lrt <- glmLRT(fit)

de <- decideTestsDGE(lrt)
summary(de)

plotMD(lrt)
abline(h=c(-1, 1), col="blue")

result_paired <- as.data.frame(topTags(lrt, n = nrow(rawcounts)))
write.table(result_paired, file = "edgeR_Tuch_Tumor_vs_NonTumor_paired.csv", sep="\t", row.names = FALSE)
```


