---
title: "edgeR with GLMs"
author: Daniel Neves
date: "April 5, 2018"
output: html_document
---

Here we demonstrate the use of *edgeR* to perform a differential expression analysis using data from Tuch *et al.* ([PLOS](http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0009317)) as detailed in the edgeR manual.

The data is a set of RNA-seq samples of oral squamous cell carcinomas and matched normal tissue from three patients that were previously quantified into raw counts. 

We will first use R to explore general charasteristics of this datasets, and then use *edgeR* to do a differential expression analysis of Tumor vs Non-Tumor samples. We will start with a simple pairwise comparison of the Tumor and Non-Tumor samples, and then repeat the analysis adding the patient pairing information to the model design.

# Setting up the environment

First we need to make sure that R is running on the same directory where we placed the counts file (edgeR_example1_Tuch.tab). To do this either type `setwd("path/to/directory")` in the R console, or navigate to the counts directory using the *Files* panel in RStudio and select "Set As Working Directory".

![](images/setwd.png)

# Load the count data

We start by importing the counts table into R using the `read.delim` function. Other functions to import tables include `read.table` and `read.csv`. We also specify that the values in the tables are separated by a TAB. You can type `?read.delim` in the R console to display the documentation of the function.

```{r}
rawdata <- read.delim("edgeR_example1_Tuch.tab", sep = "\t")
```

To check that the data was loaded properly we can use functions such as `head` (to displays the first lines of the table), `dim` (to display the dimensions of the table) and `summary` (to display summary statistics for each column). In RStudio you can also type `View(rawdata)` to view the full table on a separate window.

```{r}
head(rawdata)
dim(rawdata)
summary(rawdata)
```

# Exploring general characteristics of the raw data

Before we go further in the analysis, it's good practice to inspect the general characteristics of the data. This will allow us to make informed choices on further analysis steps, as well as prevent mistakes.

For convenience, we separate the table in two: one containing the counts for all samples (columns 2 to 7), and another containing only the list of gene names (column 1).

```{r}
rawcounts <- rawdata[, 2:7]
genes <- rawdata[, 1]
```

The next expression checks for missing values is the count data. There are no missing values.

```{r}
all(is.na(rawcounts) == FALSE)
```

We then inspect the total number of reads in each sample (or column). We can do this with the `colSums` function. We can also pass these totals to the barplot function to quickly visualize them. 

```{r}
colSums(rawcounts)

barplot(colSums(rawcounts), ylab="Total number of reads", xlab="Sample ID")
```

`ggplot2` is a popular graphics library for R that allows us to make more complex visualizations. Here we plot the distributions of counts in each sample as boxplots, and as density plots. *Note that ggplot2 warns us that some values could not represented in the plot, due to the logarithmic transformation.*

```{r}
library(ggplot2)
library(reshape2)

ggplot(melt(rawcounts, measure.vars = 1:6), aes(y=value, x=variable, col=variable)) + 
  geom_boxplot() +                                    # we want a boxplot
  scale_y_log10()                                     # y scale is log10

ggplot(melt(rawcounts, measure.vars = 1:6), aes(x=value, col=variable)) + 
  geom_density() +                                    # we want a density plot
  scale_x_log10()                                     # x scale is log10
```

We can also inspect sample-to-sample correlations using the `cor` function. This will produce a matrix correlations for all pairs of samples. By default, `cor` calculates the Pearson correlation coefficient between samples. Often, it is useful to compare this with a more robust metric, such as the rank-based Spearman correlation.

Additionally, when working with gene expression data, it is better to first transform the raw counts to log counts, to avoid the correlations being artificially determined by just a few very expressed genes.

```{r}
log10_rawcounts <- log10(rawcounts + 1)

cor(log10_rawcounts, method="pearson")
```

A useful visualization here is to display this correlation matrix as a heatmap. Here we display heatmaps of Pearson and Spearman correlations using the built-in `heatmap` function. The `heatmap` function will also apply hierarquical clustering to the matrix in order to group together the more similar samples.

```{r}
heatmap(as.matrix(cor(log10_rawcounts, method="pearson")), 
        main="Clustering of Pearson correlations", scale="none")
heatmap(as.matrix(cor(log10_rawcounts, method="spearman")), 
        main="Clustering of Spearman correlations", scale="none")
```

# Simple pairwise differential expression analysis with edgeR 

We need to import edgeR into the R environment.

```{r}
library(edgeR)
```

We start by telling edgeR where our raw counts are, and calculate normalization factors.

```{r}
y <- DGEList(counts=rawcounts, genes=genes)
y <- calcNormFactors(y)

y$samples
```

We can get the normalized counts using the `cpm` function. Compare the distributions of these CPM with the ones we generated from the raw counts.

```{r}
cpms <- as.data.frame(cpm(y, normalized.lib.sizes = TRUE))

ggplot(melt(cpms, measure.vars = 1:6), aes(x=value, col=variable)) + 
  geom_density() +                                    # we want a density plot
  scale_x_log10()                                     # x scale is log10

ggplot(melt(cpms, measure.vars = 1:6), aes(y=value, x=variable, col=variable)) + 
  geom_boxplot() +                                    # we want a boxplot
  scale_y_log10()                                     # y scale is log10
```

After normalization, we can now produce a Multidimensional Scaling Plot (MDS) using the function `plotMDS`. This visualization places the samples on a plane such that more similar samples appear closer together. We can immediately notice that the tumor samples are separated from the non-tumor samples on the first component (x axis), and that they appear to display more variability than the non-tumor samples.

```{r}
plotMDS(y)
```

We now define the design of our comparison. We want to compare Tumor to Non-Tumor samples.

```{r}
Tissue <- factor(c("N","T","N","T","N","T"))

design <- model.matrix(~Tissue)
rownames(design) <- colnames(y)

design
```

Next we use this design to conduct the test of differential expression. In *edgeR*, this is done in 3 steps: estimation of the negative binomial dispersions (`estimateDisp`), fitting of the negative binomial model to the count data (`glmFit`) and hypothesis testing (`glmLRT`).

```{r}
y <- estimateDisp(y, design, robust=TRUE)
fit <- glmFit(y, design)
lrt <- glmLRT(fit)
```

We now check how many genes were differentially expressed.

```{r}
summary(decideTestsDGE(lrt))
```

A common visualization of differential expression results is *MA-plot* that displays the relationship between a genes' mean expression and its fold-change between experimental conditions. In *edgeR* this is done with the `plotMD` function. Up-regulated genes are indicated in red, and down-regulated genes are indicated in blue. The horizontal lines indicate 2x fold-changes.

```{r}
plotMD(lrt)
abline(h=c(-1, 1), col="blue")
```

We can retrieve a table with all the results of differential expression using the `topTags` function. We also save it to a file so we can latter open it in Excel.

```{r}
result <- as.data.frame(topTags(lrt, n = nrow(rawcounts)))

head(result)

write.table(result, file = "edgeR_Tuch_Tumor_vs_NonTumor.csv", sep="\t", row.names = FALSE)
```

# A more complex design: adding patient pairing information

Recall that tumor and non-samples were collected from 3 patients. Until now we have ignored this information in our design. Here we repeat the analysis by adding the sample pairing information to our model design, that will allow us to adjust for differences between patients.

*Note that we only need to change our design definition. The rest of the commands are exactly the same as above.*

```{r}
Patient <- factor(c(8, 8, 33, 33, 51, 51))
Tissue <- factor(c("N","T","N","T","N","T"))

design <- model.matrix(~Patient + Tissue)
rownames(design) <- colnames(y)

design

y <- DGEList(counts=rawcounts, genes=genes)
y <- calcNormFactors(y)
y <- estimateDisp(y, design, robust=TRUE)
fit <- glmFit(y, design)
lrt <- glmLRT(fit)

de <- decideTestsDGE(lrt)
summary(de)

plotMD(lrt)
abline(h=c(-1, 1), col="blue")

result_paired <- as.data.frame(topTags(lrt, n = nrow(rawcounts)))
write.table(result_paired, file = "edgeR_Tuch_Tumor_vs_NonTumor_paired.csv", sep="\t", row.names = FALSE)
```

# Putting it all together

To summarize, the workflow for differential expression analysis in R using *edgeR* is usually comprised of the following steps:

1. Loading the count data (`?read.delim`).
2. Inspection of the count data through summarization and visualization (`?summary`, `?plot`, ...).
3. Initialization of the *edgeR* structure (`?DGEList`)
4. Normalization of the count data to CPM (`?calcNormFactors`).
5. Definition of the design matrix (`?model.matrix`).
6. Estimation of the negative binomial dispersions (`?estimateDisp`).
7. Fitting of the negative binomial model to the count data (`?glmFit`).
8. Hypothesis testing (`?glmLRT`).
9. Obtaining the table with results and saving to a file (`?topTags`).


